### PR TITLE
[Java] Support concurrent actor calls API.

### PR DESCRIPTION
**Pull request by jovany-wang**
## Why are these changes needed?

This PR implements concurrent actor calls in Java.

Some hack changes:
    - Attach native thread to JVM to fetch jvm env.
    - Add a map for fetching the corresponding Java task executor of the thread.
  
## Related issue number
Python implementation is #6053

### CODE REVIEW

File: java/runtime/src/main/java/org/ray/runtime/MultiRuntimeHelper.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
+package org.ray.runtime;
+
+import org.ray.api.id.UniqueId;
+import org.ray.runtime.task.TaskExecutor;
+import java.util.HashMap;
+
+public class MultiRuntimeHelper {
```
Comment (zhijunfu) [inline java/runtime/src/main/java/org/ray/runtime/MultiRuntimeHelper.java]: MultiWorkerRuntimeHelper? 

This file can be renamed as well.

File: java/runtime/src/main/java/org/ray/runtime/runner/RunManager.java [hunk 2]
```diff
@@ -233,6 +233,11 @@ private String startRedisInstance(String ip, int port, String password, Integer
 
         String name = shard == null ? "redis" : "redis-" + shard;
         startProcess(command, null, name);
+        try {
+          TimeUnit.SECONDS.sleep(2);
+        } catch (InterruptedException e) {
+
+        }
```
Comment (zhijunfu) [inline java/runtime/src/main/java/org/ray/runtime/runner/RunManager.java]: what is this for?

File: java/runtime/src/main/java/org/ray/runtime/util/FileUtil.java [hunk 3]
```diff
@@ -34,9 +34,9 @@ public File getFile() {
 
     @Override
     public void close() {
-      if (!file.delete()) {
-        LOGGER.warn("Couldn't delete temp file {}", file.getAbsolutePath());
-      }
+//      if (!file.delete()) {
+//        LOGGER.warn("Couldn't delete temp file {}", file.getAbsolutePath());
+//      }
```
Comment (zhijunfu) [inline java/runtime/src/main/java/org/ray/runtime/util/FileUtil.java]: should these be removed?

File: java/test/src/main/java/org/ray/api/test/ActorConcurrentCallsTest.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+
+@Test(groups = {"directCall"})
+public class ActorConcurrentCallsTest extends BaseTest {
```
Comment (zhijunfu) [inline java/test/src/main/java/org/ray/api/test/ActorConcurrentCallsTest.java]:
```suggestion
public class ActorConcurrentCallTest extends BaseTest {
```

File: src/ray/core_worker/lib/java/org_ray_runtime_RayNativeRuntime.cc [hunk 5]
```diff
@@ -6,7 +6,7 @@
 #include "ray/core_worker/lib/java/jni_utils.h"
 
 thread_local JNIEnv *local_env = nullptr;
-thread_local jobject local_java_task_executor = nullptr;
+//thread_local jobject local_java_task_executor = nullptr;
```
Comment (zhijunfu) [inline src/ray/core_worker/lib/java/org_ray_runtime_RayNativeRuntime.cc]: ?

File: src/ray/core_worker/lib/java/org_ray_runtime_RayNativeRuntime.cc [hunk 6]
```diff
@@ -89,11 +103,11 @@ JNIEXPORT jlong JNICALL Java_org_ray_runtime_RayNativeRuntime_nativeInitCoreWork
 JNIEXPORT void JNICALL Java_org_ray_runtime_RayNativeRuntime_nativeRunTaskExecutor(
     JNIEnv *env, jclass o, jlong nativeCoreWorkerPointer, jobject javaTaskExecutor) {
   local_env = env;
-  local_java_task_executor = javaTaskExecutor;
+//  local_java_task_executor = javaTaskExecutor;
   auto core_worker = reinterpret_cast<ray::CoreWorker *>(nativeCoreWorkerPointer);
   core_worker->StartExecutingTasks();
   local_env = nullptr;
-  local_java_task_executor = nullptr;
+//  local_java_task_executor = nullptr;
```
Comment (zhijunfu) [inline src/ray/core_worker/lib/java/org_ray_runtime_RayNativeRuntime.cc]: ?

File: src/ray/core_worker/lib/java/org_ray_runtime_RayNativeRuntime.cc [hunk 7]
```diff
@@ -39,9 +39,23 @@ JNIEXPORT jlong JNICALL Java_org_ray_runtime_RayNativeRuntime_nativeInitCoreWork
...
-         std::vector<std::shared_ptr<ray::RayObject>> *results) {
+         std::vector<std::shared_ptr<ray::RayObject>> *results,
+         const ray::WorkerID &worker_id) {
         JNIEnv *env = local_env;
+        if (!env) {
+          // Attach the native thread to JVM.
+          auto status =
+              jvm->AttachCurrentThreadAsDaemon(reinterpret_cast<void **>(&env), NULL);
+          RAY_CHECK(status == JNI_OK) << "Failed to get JNIEnv. Return code: " << status;
+          local_env = env;
```
Comment (zhijunfu) [inline src/ray/core_worker/lib/java/org_ray_runtime_RayNativeRuntime.cc]: why is this required?

File: java/test/src/main/java/org/ray/api/test/ActorConcurrentCallsTest.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+        public int compare(Integer o1, Integer o2) {
+          return o1.compareTo(o2);
+        }
+      });
+      return li;
+    }
+
+  }
+
+  public void testConcurrentCalls() {
```
Comment (zhijunfu) [inline java/test/src/main/java/org/ray/api/test/ActorConcurrentCallsTest.java]: could you add some comments on what this is testing for?

Comment (zhijunfu) [reply java/test/src/main/java/org/ray/api/test/ActorConcurrentCallsTest.java]: and it might make sense to add some test to verify the methods are indeed executed in parallel.

File: java/runtime/src/main/java/org/ray/runtime/MultiRuntimeHelper.java (hunk 1 above)
Comment (zhijunfu) [reply java/runtime/src/main/java/org/ray/runtime/MultiRuntimeHelper.java]: also the name `helper` itself seems a bit more generic than what it actually does.

File: java/runtime/src/main/java/org/ray/runtime/RayMultiWorkerNativeRuntime.java [hunk 9]
```diff
@@ -65,6 +65,7 @@ public RayMultiWorkerNativeRuntime(RayConfig rayConfig, FunctionManager function
       threads[i] = new Thread(() -> {
         RayNativeRuntime runtime = new RayNativeRuntime(rayConfig, functionManager);
         runtimes[workerIndex] = runtime;
+        MultiRuntimeHelper.put(runtime.getWorkerContext().getCurrentWorkerId(), runtime.getTaskExecutor());
```
Comment (zhijunfu) [inline java/runtime/src/main/java/org/ray/runtime/RayMultiWorkerNativeRuntime.java]: is it possible to put this into the thread local runtime?

File: java/runtime/src/main/java/org/ray/runtime/runner/RunManager.java (hunk 2 above)
Comment (jovany-wang) [reply java/runtime/src/main/java/org/ray/runtime/runner/RunManager.java]: unintentional change~~~

File: java/runtime/src/main/java/org/ray/runtime/util/FileUtil.java (hunk 3 above)
Comment (jovany-wang) [reply java/runtime/src/main/java/org/ray/runtime/util/FileUtil.java]: It's used to test on my mbp.

File: src/ray/core_worker/lib/java/org_ray_runtime_RayNativeRuntime.cc (hunk 7 above)
Comment (jovany-wang) [reply src/ray/core_worker/lib/java/org_ray_runtime_RayNativeRuntime.cc]: `env` is thread local now. 
It's null in the concurrent thread pool.

File: java/test/src/main/java/org/ray/api/test/ActorConcurrentCallsTest.java (hunk 8 above)
Comment (jovany-wang) [reply java/test/src/main/java/org/ray/api/test/ActorConcurrentCallsTest.java]: It's actually that the case must executed in parallel because the first task is pending until other tasks putting items into the queue.

File: java/test/src/main/java/org/ray/api/test/ActorConcurrentCallsTest.java (hunk 4 above)
Comment (jovany-wang) [reply java/test/src/main/java/org/ray/api/test/ActorConcurrentCallsTest.java]: done

File: java/runtime/src/main/java/org/ray/runtime/RayMultiWorkerNativeRuntime.java (hunk 9 above)
Comment (jovany-wang) [reply java/runtime/src/main/java/org/ray/runtime/RayMultiWorkerNativeRuntime.java]: Because of the native thread pool, it seems hard to put it into thread local runtime.
This helper is used to cache the worker id -> runtime/executor.

File: java/runtime/src/main/java/org/ray/runtime/MultiRuntimeHelper.java [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+
+  private static HashMap<UniqueId, TaskExecutor> taskExecutors = new HashMap<>();
```
Comment (kfstorm) [inline java/runtime/src/main/java/org/ray/runtime/MultiRuntimeHelper.java]: `ConcurrentHashMap`

Comment (kfstorm) [inline java/runtime/src/main/java/org/ray/runtime/MultiRuntimeHelper.java]: It's better to put the map into `RayMultiWorkerNativeRuntime`. In the `get` method, we can find the runtime instance with `((RayMultiWorkerNativeRuntime)Ray.internal()).getRuntime(workerId)`. Basically, the main purpose of making the `get` method static is that you don't need to get a reference of the `MultiRuntimeHelper` instance in JNI.

File: java/runtime/src/main/java/org/ray/runtime/MultiRuntimeHelper.java (hunk 1 above)
Comment (raulchen) [reply java/runtime/src/main/java/org/ray/runtime/MultiRuntimeHelper.java]: how about just add a static `getTaskExecutorByWorker` in `TaskExecutor`? 
(we can set the mapping in `TaskExecutor`'s constructor.

File: java/runtime/src/main/java/org/ray/runtime/MultiRuntimeHelper.java (hunk 10 above)
Comment (raulchen) [inline java/runtime/src/main/java/org/ray/runtime/MultiRuntimeHelper.java]: can change to `ConcurrentHashMap` and remove `synchronized`

File: src/ray/core_worker/lib/java/org_ray_runtime_RayNativeRuntime.cc [hunk 11]
```diff
@@ -89,11 +102,9 @@ JNIEXPORT jlong JNICALL Java_org_ray_runtime_RayNativeRuntime_nativeInitCoreWork
 JNIEXPORT void JNICALL Java_org_ray_runtime_RayNativeRuntime_nativeRunTaskExecutor(
     JNIEnv *env, jclass o, jlong nativeCoreWorkerPointer, jobject javaTaskExecutor) {
   local_env = env;
-  local_java_task_executor = javaTaskExecutor;
```
Comment (zhijunfu) [inline src/ray/core_worker/lib/java/org_ray_runtime_RayNativeRuntime.cc]: seems `javaTaskExecutor` is unused now?

File: java/runtime/src/main/java/org/ray/runtime/MultiRuntimeHelper.java (hunk 1 above)
Comment (jovany-wang) [reply java/runtime/src/main/java/org/ray/runtime/MultiRuntimeHelper.java]: Thanks. I put these methods into `TaskExecutor`

Comment (jovany-wang) [reply java/runtime/src/main/java/org/ray/runtime/MultiRuntimeHelper.java]: Thanks. I put these methods into TaskExecutor.

File: java/api/src/main/java/org/ray/api/options/ActorCreationOptions.java [hunk 12]
```diff
@@ -62,8 +66,18 @@ public Builder setJvmOptions(String jvmOptions) {
       return this;
     }
 
+    public Builder setMaxConcurrency(int maxConcurrency) {
```
Comment (raulchen) [inline java/api/src/main/java/org/ray/api/options/ActorCreationOptions.java]: document this. you can just follow python's documentation

File: java/test/src/main/java/org/ray/api/test/ActorConcurrentCallTest.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      }
+    }
+  }
+
+  public void testConcurrentCall() {
+    TestUtils.skipTestIfDirectActorCallDisabled();
+    /// The logic of this case is that the first task will be pending until
+    /// the following 2 tasks get executed.
```
Comment (raulchen) [inline java/test/src/main/java/org/ray/api/test/ActorConcurrentCallTest.java]: this comment is out-dated?

**Review by raulchen — APPROVED** [review]
